### ISSUE TITLE
[ML] Data frame silently fails to index if aggs return a string

### ISSUE THREAD

**Issue by sophiec20** [elastic/elasticsearch#43194]
Found in 7.2.0-BC5

The following data frame preview contains a typo in `responsetime`. Composite aggs returns `avg(resposetime)` as a string `"NaN"` . 

```
GET _data_frame/transforms/_preview
{
  "id": "farequote-airline-100",
  "source" : { "index": [ "farequote-*" ] },
  "dest"     : {  "index": "df-farequote-airline-100" },
  "pivot": {
    "group_by": {
      "airline": { "terms": { "field": "airline" }}
    },
    "aggregations": {
      "last_updated": { "max": { "field": "@timestamp" } },
      "mean_response": { "avg": {  "field": "resposetime" } },
      "count": { "value_count": { "field": "airline" } }
    }
  }
}
```
returns
```
{
  "preview" : [
    {
      "last_updated" : "2019-02-11T23:59:02.000Z",
      "count" : 8728.0,
      "mean_response" : "NaN",
      "airline" : "AAL"
    },
...
```

The data frame silently fails to index these values resulting in an entirely empty dest index. 

However there are no errors in elasticsearch logs. There are no warnings or errors in data frame notifications, and the stats make it appear that the data frame transform was successful.

```
{
  "count" : 1,
  "transforms" : [
    {
      "id" : "farequote-airline-100",
      "state" : {
        "task_state" : "stopped",
        "indexer_state" : "stopped",
        "checkpoint" : 1,
        "progress" : {
          "total_docs" : 86274,
          "docs_remaining" : 0,
          "percent_complete" : 100.0
        }
      },
      "stats" : {
        "pages_processed" : 2,
        "documents_processed" : 86274,
        "documents_indexed" : 19,
        "trigger_count" : 1,
        "index_time_in_ms" : 136,
        "index_total" : 1,
        "index_failures" : 0,
        "search_time_in_ms" : 25,
        "search_total" : 2,
        "search_failures" : 0
      },
      "checkpointing" : {
        "operations_behind" : 0
      }
    }
  ]
}
```

We require better handling for these errors under conditions such as when:
1) something like a typo causes complete failure of the data frame (note, if the other aggs worked, perhaps this should have completed without `resposetime`)
2) aggregations may sporadically return strings (`NaN` or `Infinity`) .. these should not silently fail. Is it possible to keep counts of these as `index_failures` or other stat?

Messages should additionally be logged to data frame notifications, where appropriate, to ease operational management.

**Comment by davidkyle** [comment]
> The data frame silently fails to index these values resulting in an entirely empty dest index.

I can't reproduce this on master or the 7.2 bc6. I can see documents written to the index with a `null` value for the `mean_respose` field:

```
      "_source" : {
          "last_updated" : null,
          "count" : 8728.0,
          "mean_response" : null,
          "airline" : "AAL"
        }
```

**Comment by sophiec20** [comment]
I'll retry on BC6.

Note that we have also seen silent failures when the deduced mapping was `half_float` and the `sum(value)` overflowed. cc @benwtrent

**Comment by benwtrent** [comment]
@sophiec20 yes, 7.3 is now writing an audit when there are bulk indexing issues. If desired, it can be backported to 7.2. However, I am not sure that is necessary because the UI is not displaying audit messages yet.

**Comment by sophiec20** [comment]
@benwtrent Not worth a backport to 7.2 at this time. Do these failures get counted by `_stats`?

**Comment by benwtrent** [comment]
@sophiec20 no, it will only write a `warning` log entry.

**Comment by benwtrent** [comment]
For the sum specific issue: https://github.com/elastic/elasticsearch/pull/43213
Alerting when there are bulk index issues: https://github.com/elastic/elasticsearch/pull/43106
Handling `null` values better for aggs: https://github.com/elastic/elasticsearch/pull/42966

@sophiec20 and @davidkyle I think these three PRs address the concerns in this issue.

**Comment by sophiec20** [comment]
@benwtrent yes, those 3 PRs do address many of the concerns around silently failing..

Before we close this I have one more question, under what conditions does stats `index_failures` get incremented?

**Comment by benwtrent** [comment]
`index_failures` gets incremented if there is an exception thrown during the bulk indexing process. I am not 100% sure why this type of behavior is the way it is, but the Rollups folks may have better input as to why we don't count a bulk indexing failure as an `index_failure` (it seems to me like we should).

**Comment by benwtrent** [comment]
@polyfractal @jimczi what do y'all think? Should BulkIndex failures increment the `index_failure` statistic for the `AsyncTwoPhaseIndexer`? It seems to me like they should. Otherwise there is no obvious indication that such a failure happened.

**Comment by jimczi** [comment]
> Should BulkIndex failures increment the index_failure statistic for the AsyncTwoPhaseIndexer? 

I am not sure because it currently contains the number of times that the job failed due to a failure in bulk indexing. If the entire bulk request fails we stop the job and increment the counter whereas if only individual documents failed we ignore and continue the processing. So maybe we should treat all indexing errors as fatal and stop the job ? Most of these errors should be transient so resuming a job a with a backoff policy should be enough. For rollups it is acceptable to loose some documents (debatable but we've made this decision long time ago ;)) but there is no way to go back in time to replay those missing documents so I'd prefer to fail rather than incrementing a counter and move to indexer state.

**Comment by sophiec20** [comment]
> ... it is acceptable to loose some documents ...

It depends, yes and no, depending on the use case. Being able to quantify how many documents are lost in bulk allows the user to somewhat quantify this impact and they can make a choice. I don't recall being able to easily tell how many documents were lost in `_bulk` .. can we tell?

**Comment by jimczi** [comment]
> I don't recall being able to easily tell how many documents were lost in _bulk .. can we tell?

Yes that's what we ignore here. The bulk response contains the list of success and failures so we can increment the counter of missing documents easily. My point here is that we don't have a way to tell a job to retry from a specific point so if you have missing documents and you're not ok with it the only solution currently is to remove the job and start again.

**Comment by polyfractal** [comment]
> I don't recall being able to easily tell how many documents were lost in _bulk .. can we tell?

There's not a specific counter, no.  The bulk response has a high-level property that tells you if there are failures.  But if you want the total count and how they failed, you have to iterate over the response and count/collect the individual failures.

> So maybe we should treat all indexing errors as fatal and stop the job ?  [...] but there is no way to go back in time to replay those missing documents so I'd prefer to fail rather than incrementing a counter and move to indexer state.

Yeah I'd be fine with failing the current indexer run if there are bulk failures, which will force the indexer to backtrack and re-run on the next trigger.  There's a `todo` note in the bulk section which basically says "what should we do here?" :)

Another solution is to re-index the failed documents before continuing.  We didn't implement that because it opens up a whole host of non-trivial edge-cases (what happens if those fail too? How do you chain it into the existing indexer cycle? etc etc).

> Handling null values better for aggs: #42966

Just an FYI, we [implemented some "inspection helpers"](
https://github.com/elastic/elasticsearch/pull/36020) for SQL a while ago.  Given an internal agg result, the helpers will tell you if they have a value, using the specific rules for each type of agg (`min` uses `+Infinity`, etc).  Probably the same as what you implemented, just maintained by the agg framework so you don't have to :)

**Comment by benwtrent** [comment]
Awesome, I (or somebody else) will write up a PR so that we raise when there is a bulk index failure. This will simplify other aspects. 

I will in parallel look into the `inspection helpers`. I was unsure how to get to the `InternalAggregation` class from the `Aggregation` value in the search response. Will dig around to see what I find. 



I do not think `AsyncTwoPhaseIndexer` should auto-retry. I do not want to stick all users and future users of the class with a sub-optimal abstraction.

**Comment by hendrikmuhs** [comment]
I created #44101 as a follow up regarding behavior on bulk index failures.

Best to my knowledge this issue can be closed now as the original issue has been solved and #44101 covers the rest.

@benwtrent please re-open if I misunderstood something
